pluma/test/shelltest.py

```
@@ -52,41 +55,87 @@ def test_body(self):
+        if self.runs_in_shell:
+            output = CommandRunner.run(test_name=self._test_name, console=console,
+                                       command=script, timeout=self.timeout)
+        else:
+            output = CommandRunner.run_raw(test_name=self._test_name, console=console,
+                                           command=script, timeout=self.timeout)
+
+        if self.should_print or self.should_not_print:
+            CommandRunner.check_output(test_name=self._test_name, command=script, output=output,
+                                       should_print=self.should_print,
+                                       should_not_print=self.should_not_print)
 
 
 class CommandRunner():
     @staticmethod
-    def run(test_name: str, console: ConsoleBase, command: str,
-            timeout: int = None, should_print: list = None, should_not_print: list = None, runs_in_shell: bool = True):
-        should_print = should_print or []
-        should_not_print = should_not_print or []
-        log = Logger()
+    def run(test_name: str, console: ConsoleBase, command: str, timeout: int = None):
+        '''Run a command in a Shell context'''
+        retcode_token = 'pluma-retcode='
+        command += " && echo pluma-retcode=$?"
+        output, matched = console.send_and_expect(
+            command, timeout=10, match=retcode_token+r'-?\d+')
+
+        if not matched:
+            CommandRunner.log_error(test_name=test_name, sent=command, output=output,
+                                    error='No response within timeout, or failed device'
+                                    ' failed to send return code. If this is not running'
+                                    ' in a shell, set "runs_in_shell" to "false".')
+
+        output = re.sub(re.escape(matched) + r'$', '', output)
+        retcode_match = re.search(retcode_token+r'(-?\d+)', matched)
+        if not retcode_match:
+            raise Exception('Failed to find return code value')
```

Would maybe be better to raise a named exception.
E.g.
https://github.com/Witekio/pluma-automation/blob/22c60cf27ee75114523d66f973123b3e42f20fa6/pluma/test/test.py#L51

---

Done